### PR TITLE
Poseidon version of SnapDeals

### PR DESCRIPTION
**Pull request by Kubuxu**
Signed-off-by: Jakub Sztandera <kubuxu@protocol.ai>

### CODE REVIEW

**Review by DrPeterVanNostrand — COMMENTED** [review]
Thanks for writing this! Looks good to me.

A couple of comments/questions:
- It looks like this is a single partition circuit, am I correct in thinking that?
- I think that restructuring this module makes sense when the vanilla side is built (mostly Merkle tree/proof shape changes) after the trusted-setup is run, e.g. maybe splitting the `empty_sector_update` crate into two modules `sha` and `poseidon`.

**Comment by Kubuxu** [comment]
> t looks like this is a single partition circuit, am I correct in thinking that?

Yes, it is a single partition.

> I think that restructuring this module makes sense when the vanilla side is built (mostly Merkle tree/proof shape changes) after the trusted-setup is run, e.g. maybe splitting the empty_sector_update crate into two modules sha and poseidon (which signify TreeD's hasher and shape).

Agreed, I wanted to limit changes to SHA version that is why I moved new vanilla definitions to the circuit module itself and avoided touching the SHA sides in general where possible.

File: storage-proofs-update/src/challenges.rs
Comment (Kubuxu) [reply storage-proofs-update/src/challenges.rs:77]: Yes.

File: storage-proofs-update/src/vanilla.rs
Comment (Kubuxu) [reply storage-proofs-update/src/vanilla.rs:477]: Will change to `D`, TreeDDomain would conflict but I can do it as well.

File: storage-proofs-update/tests/common/mod.rs
Comment (Kubuxu) [reply storage-proofs-update/tests/common/mod.rs:1]: I tried but I couldn't import it, I think it was being recognised as a test itself.

File: storage-proofs-update/tests/common/mod.rs
Comment (Kubuxu) [reply storage-proofs-update/tests/common/mod.rs:120]: The compound had its own version, I can check if they are equivalent.

File: storage-proofs-update/src/circuit_poseidon.rs
Comment (Kubuxu) [reply storage-proofs-update/src/circuit_poseidon.rs:27]: I wanted to keep same PublicParams for consistency in the future when both version of the circuit might be verified by the same endpoint.

File: storage-proofs-update/src/circuit_poseidon.rs
Comment (Kubuxu) [reply storage-proofs-update/src/circuit_poseidon.rs:364]: I tried ignoring them but I couldn't figure it out for some reason, re PublicParams see above, I wanted to keep the same PublicParams, which could be modified to better suite both circuit but I also wanted to avoid affecting SHA circuit in any way right now.

File: storage-proofs-update/tests/common/mod.rs
Comment (Kubuxu) [reply storage-proofs-update/tests/common/mod.rs:1]: Now it worked.

File: storage-proofs-update/src/circuit_poseidon.rs
Comment (Kubuxu) [reply storage-proofs-update/src/circuit_poseidon.rs:364]: fixed

**Comment by Kubuxu** [comment]
For some reason clippy is not happy:
```
error[E0310]: the parameter type `U` may not live long enough
   --> storage-proofs-core/tests/por_circuit.rs:393:16
    |
375 | fn test_por_no_challenge_input<U, V, W>(sector_nodes: usize)
    |                                - help: consider adding an explicit lifetime bound...: `U: 'static`
...
393 |     let tree = create_tree::<DiskTree<PoseidonHasher, U, V, W>>(&leafs, tmp_path);
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...so that the type `U` will meet its required lifetime bounds

error[E0310]: the parameter type `V` may not live long enough
   --> storage-proofs-core/tests/por_circuit.rs:393:16
    |
375 | fn test_por_no_challenge_input<U, V, W>(sector_nodes: usize)
    |                                   - help: consider adding an explicit lifetime bound...: `V: 'static`
...
393 |     let tree = create_tree::<DiskTree<PoseidonHasher, U, V, W>>(&leafs, tmp_path);
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...so that the type `V` will meet its required lifetime bounds

error[E0310]: the parameter type `W` may not live long enough
   --> storage-proofs-core/tests/por_circuit.rs:393:16
    |
375 | fn test_por_no_challenge_input<U, V, W>(sector_nodes: usize)
    |                                      - help: consider adding an explicit lifetime bound...: `W: 'static`
...
393 |     let tree = create_tree::<DiskTree<PoseidonHasher, U, V, W>>(&leafs, tmp_path);
    |                ^^^^^^^^^^^^^^^
```

**Review by DrPeterVanNostrand — APPROVED** [review]
Looks good to me. Thanks!

**Comment by cryptonemo** [comment]
@Kubuxu Can we get this rebased against latest so that it can have final review before merge?

File: storage-proofs-update/tests/circuit_poseidon.rs
Comment (cryptonemo) [inline storage-proofs-update/tests/circuit_poseidon.rs:133]: I'll have to check what CI is doing, but since there's already a test with this name using the `isolated-testing` flag, I would think that both tests are getting run at the same time, which may result in CI RAM issues.  Once verified, we may need to update the test names and add entries to run them in the .circleci/config

Comment (cryptonemo) [reply storage-proofs-update/tests/circuit_poseidon.rs:133]: Ah, nice, since they are separate tests, they appear to also be run independently on CI.

**Review by cryptonemo — APPROVED** [review]

**Comment by BlocksOnAChain** [comment]
@Kubuxu @DrPeterVanNostrand @cryptonemo   - Just wanted to say I'm glad we got this one in the scope. Great idea and all the work behind it. 🎉 👍

**Pull request merged**
